rfd/XXXX-teleport-release-engineering.md

```diff
@@ -0,0 +1,416 @@
+    - Ephemeral AWS EC2 arm64 boxes for arm64 linux builds. Pool has a max of 10
+    instances
+    - Ephemeral EC2 instance with a persistent volume for building `deb` 
+    and `rpm` packages.
+    
+### Goals of a Release Process
+
+At a a high level, we want a release process that is:  
+  * *Consistent* - we want to provide a stable framwork for delivering 
+    releases reliably from source,
+  * *Agile* - we want to be able to quickly integrate new delvery channels
+    or products without having to redesign everything. 
+  * *Identifiable* - we can identify all the tools and components used
+    to deliver a release, in order to mitigate supply chain attacks
+
+We also want some functional requirements:
+
+ 1. An engineer *must* be able to at least construct the `teleport`, `tsh`, and `tctl` binaries
+    for their current operating system with a minimum of fuss.
+ 2. An engineer *must* be able to run close to 100% of all tests at their desk
+    with a minimum of fuss 
+ 3. The release envirionment should be as close as practical to the CI 
+    environment, ideally the same. We don't want surprises come release time. 
+ 4. An engineer should be able to cut a debug or testing release at any time 
+    using the proper release procedure and tooling. Especially if that release
+    will travel outside the Teleport organisation.
+
+Non-functional requirements:
+
+### Suggestions
+
+#### **Draw some boundaries**
+
+One cause of the complexity in our build system is that we have not (in the
+last 2 years at least) collectively sat down and thought about what we want
+our Delivery system to *be*. What is its *purpose*, and what boundaries can
+we put on it.
+
+Some open questions are:
+
+ - What tasks should an engineer be able to do at their desk?
+   - Build binaries
+   - Run Tests
+ - Where is the boundary between
+   - Desk/CI
+   - Desk/Delivery
+   - Using make vs Go vs GHA workflow (see below)
+ - How much should we exercise the release system in CI?
+
+#### **Buildbox management**
+
+Whole separate RFD. General principles are:
+
+  1. Stop rebuilding buildbox on every build. Rebuild on change to buildbox 
+     content.
+  2. All buildbox revisions are immutable and versioned. 
+  3. All build automation must refer to expected buildbox unambiguously, so
+     that rebuilds can happen in the context of the original build, not
+     something that may ahve been subsequently altered.
+  4. Build scripts must opt-in to updated buildboxes. No change will be thrust
+     upon a build script.
+  5. While we're dreaming, one multiarch buildbox for both CI and Delivery
+     for any given revision of Teleport
+
+#### **Prefer writing Go to any other method**
+
+If at all possible, any even slightly complex task for either CI or 
+the Delivery pipeline should be written in Go. Go is much less esoteric than
+`bash` or `make`, and less prone to subtle bugs. 
+
+Go also avoid locking us into a vendor's build ourchestration system. We have
```

Comment:
I would not say that is a property of using Go. It really is just making sure all the "build scripts" are independent of the CI system, and the CI system just calls them. Putting too much logic into the CI-specific build system makes it harder to migrate later, makes it harder to perform break-glass operations when CI is unavailable, and makes it harder to run locally to test.

Reply:
```suggestion
Capturing processes _outside_ a vendor's build orchestration system is also a useful goal. We have
```